otherlibs/systhreads/st_stubs.c

```diff
@@ -572,9 +572,12 @@ CAMLprim value caml_thread_initialize(value unit)
 CAMLprim value caml_thread_cleanup(value unit)
 {
   if (Tick_thread_running){
-    atomic_store_release(&Tick_thread_stop, 1);
+    caml_plat_lock (&Tick_thread_control.mu);
+    Tick_thread_control.state = Tick_stop;
+    caml_plat_unlock (&Tick_thread_control.mu);
+    caml_plat_signal (&Tick_thread_control.cond);
```

Comment:
```suggestion
    caml_plat_signal (&Tick_thread_control.cond);
    caml_plat_unlock (&Tick_thread_control.mu);
```
The `DEBUG_LOCK` seems to enforce that the mutex is locked when entering `caml_plat_signal`. But there is a PR currently under review which aims to remove the needless restriction.

However according to some sources online, schedulers are programmed to better handle the case where `pthread_cond_signal` is called with the mutex locked. Not an expert on these details though. Feel free to ignore this change if you have a strong opinion about it. I am curious to know the rationale for the best choice.

Reply:
I have no strong opinion and I am ready to believe that schedulers are optimized that way.